s3-native-client/src/main/java/com/amazonaws/s3/S3NativeClient.java

```diff
@@ -17,41 +19,83 @@
 import java.util.concurrent.*;
 import java.util.function.Consumer;
 
-public class S3NativeClient implements  AutoCloseable {
+public class S3NativeClient implements AutoCloseable {
     private final S3Client s3Client;
     private final String signingRegion;
 
-    public S3NativeClient(final String signingRegion,
-                          final ClientBootstrap clientBootstrap,
-                          final CredentialsProvider credentialsProvider,
-                          final long partSizeBytes,
-                          final double targetThroughputGbps) {
-        //keep signing region to construct Host header per-request
+    public S3NativeClient(final String signingRegion, final S3Client s3Client) {
         this.signingRegion = signingRegion;
-        final S3ClientOptions clientOptions = new S3ClientOptions()
-                .withClientBootstrap(clientBootstrap)
-                .withCredentialsProvider(credentialsProvider)
-                .withRegion(signingRegion)
-                .withPartSize(partSizeBytes)
+        this.s3Client = s3Client;
+    }
+
+    public S3NativeClient(final String signingRegion, final ClientBootstrap clientBootstrap,
+            final CredentialsProvider credentialsProvider, final long partSizeBytes,
+            final double targetThroughputGbps) {
+        // keep signing region to construct Host header per-request
+        this.signingRegion = signingRegion;
+        final S3ClientOptions clientOptions = new S3ClientOptions().withClientBootstrap(clientBootstrap)
+                .withCredentialsProvider(credentialsProvider).withRegion(signingRegion).withPartSize(partSizeBytes)
                 .withThroughputTargetGbps(targetThroughputGbps);
         s3Client = new S3Client(clientOptions);
     }
 
+    private void addCustomHeaders(List<HttpHeader> headers, HttpHeader[] customHeaders) {
+        assert headers != null : "Invalid argument - headers list is null";
+
+        if (customHeaders == null || customHeaders.length == 0) {
+            return;
+        }
+
+        for (HttpHeader customHeader : customHeaders) {
+
+            // Prevent duplicates and warn if any are found.
+            for (HttpHeader header : headers) {
+                if (customHeader.getName().equals(header.getName())) {
+                    Log.log(Log.LogLevel.Warn, Log.LogSubject.JavaCrtS3,
+                            "Custom header '" + customHeader.getName() + "' is overriding existing header.");
+                    headers.remove(header);
+                    break;
+                }
+            }
+
+            headers.add(customHeader);
+        }
+    }
+
+    private String getEncodedPath(String key, String customQueryParameters) {
+        String encodedPath = "/" + key;
+
+        if (customQueryParameters == null) {
+            return encodedPath;
+        }
+
+        String trimmedCustomQueryParameters = customQueryParameters.trim();
```

Comment:
Because they're already packaged as a string right now, they should already be passed in with the necessary base 64 encoding.  (Maybe there should be some better documentation there)

Reply:
I think uri encoding is the question here rather than base64.

Reply:
Also I feel like a solution that requires the caller to have done the encoding may be better, in order to avoid any of the weirdness (similar to signing) where parameters may be path-embedded and whether or not they should be encoded varies.

Reply:
Ah, my mistake!  I like Bret's thought on this.

Reply:
Looking at the code again, I now lean towards to requiring the caller to pass the encoded and combined param string. That way we don't need to worry about encoding and supporting query params with multiple values in this client